src/main/java/de/tum/in/www1/artemis/service/compass/controller/ModelIndex.java

```diff
@@ -28,44 +31,83 @@ public ModelIndex() {
     }
 
     /**
-     * Get the internal similarity id for the given model element. If the element is similar to an existing one, they share the same similarity id. Otherwise, a new id is created.
+     * Get the internal similarity ID for the given model element. If the element is similar to an existing one, they share the same similarity id, i.e. they are in the same
+     * similarity set. Otherwise, the given element does not belong to an existing similarity set and a new similarity ID is created for the element.
      *
-     * @param element a model element for which the corresponding similarity id should be retrieved
-     * @return the similarity id for the given model element
+     * @param element a model element for which the corresponding similarity ID should be retrieved
+     * @return the similarity ID for the given model element, i.e. the ID of the similarity set the element belongs to
      */
     int retrieveSimilarityId(UMLElement element) {
         if (modelElementMapping.containsKey(element)) {
             return modelElementMapping.get(element);
         }
-        // element is similar to existing element
+
+        Map<Double, Integer> similarityToIdMapping = new HashMap<>();
```

Comment:
You don't have to iterate through the list twice if you just save a reference to the current highest similarity and ID:
```suggestion
        // Pair of similarity value and similarity ID
        var bestSimilarityFit = Pair.of(-1.0, -1);
        
        for (final var knownElement : uniqueModelElementList) {
            final var similarity = knownElement.similarity(element);
            if (similarity > CompassConfiguration.EQUALITY_THRESHOLD && similarity > bestSimilarityFit.getFirst()) {
                // element is similar to existing element and has a higher similarity than another element
                bestSimilarityFit = Pair.of(similarity, knownElement.getSimilarityID());
            }
        }
```

Reply:
Good point, I applied your suggested changes.